stream: tcp: Handle retransmitted SYN with TSval

mtremer opened the pull request:
For connections that use TCP timestamps for which the first SYN packet
does not reach the server, any replies to retransmitted SYNs will be
tropped.

This is happening in StateSynSentValidateTimestamp, where the timestamp
value in a SYN-ACK packet must match the one from the SYN packet.
However, since the server never received the first SYN packet, it will
respond with an updated timestamp from any of the following SYN packets.

The timestamp value inside suricata is not being updated at any time
which should happen. This patch fixes that problem.

This problem was introduced in 9f0294fadca3dcc18c919424242a41e01f3e8318.

Signed-off-by: Michael Tremer <michael.tremer@ipfire.org>

---

This patch has been developed against the master branch, and can and should be backported as is to the 5 and 6 branches.

victorjulien commented:
Can you make sure the check all the check boxes from the PR template? Also I didn't find a record in our CLA db. Please see https://suricata.io/contribution-agreements/

Thanks!

mtremer commented:
Hello Victor,

I have sent the signed contributor agreement to the email address on the document.

Here is the rest:

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Regarding the "needs ticket" label: I am very confused about what the ways to communicate with the community. There is Redmine, there is GitHub issues, and there is GitHub pull requests. It all seems to be redundant and I could not find any hints that the project is migrating away from one to another.

Thank you for dealing with my patch.

victorjulien commented:
Thanks!

Wrt tickets, we use redmine only, not github issues. The workflow there is that normally we start with a ticket targeting the next major release (7.0rc1 in this case), and tag the ticket for 'needs backport to 5.0.x' and 'needs backport to 6.0.x' if needed. Then on our end we'll create copies of this "master" ticket and handle the backports in general. A commit message should mention the ticket `Bug: #XXXX.` so that we can easily cross check things.

Thanks again.

victorjulien commented:
Are you also able to provide a pcap or even a full Suricata-Verify test? (https://github.com/OISF/suricata-verify)

mtremer commented:
> Thanks!
> 
> Wrt tickets, we use redmine only, not github issues. The workflow there is that normally we start with a ticket targeting the next major release (7.0rc1 in this case), and tag the ticket for 'needs backport to 5.0.x' and 'needs backport to 6.0.x' if needed. Then on our end we'll create copies of this "master" ticket and handle the backports in general. A commit message should mention the ticket `Bug: #XXXX.` so that we can easily cross check things.
> 
> Thanks again.

Okay, I will know for next time.

> Are you also able to provide a pcap or even a full Suricata-Verify test? (https://github.com/OISF/suricata-verify)

I have a pcap which I cannot disclose, but anything that sends a SYN packet, a retransmitted SYN and the following SYN-ACK packet and those retries will do the trick.

I did not run suricata-verify.

victorjulien commented:
Is this the same or related to https://redmine.openinfosecfoundation.org/issues/4376 ?

mtremer commented:
Yes, the patch looks technically identical. The log message is different and at a different place.

The attached PCAP file starts with the same sequence that was triggering the bug for me.

mtremer commented:
Is there a way to re-run these? The logs have already been erased.

As far as I remember, those issues came from the build system and don't have anything to do with the patch. The patch is very simple and couldn't break the build like this.

inashivb commented:
Thank you very much for your first contribution to our project, @mtremer ! :)
> As far as I remember, those issues came from the build system and don't have anything to do with the patch. The patch is very simple and couldn't break the build like this.

Indeed. And seems like it doesn't. :) We just need a formatting fix. You could do 
```
git reset HEAD~
git add <changed file>
git clang-format
git add <changed file>
git commit
``` 
This is one of the ways to fix the formatting. The script can be found [here](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) if you do not already have it.

mtremer commented:
Okay, done. I updated the pull request and will keep my fingers crossed for the checks going through this time.

mtremer commented:
I just received an email that this has failed again; however, the website doesn't reflect that for me.

Here is what I did including the changes that I made and I pushed them to the branch. Any idea why this isn't happy?

```
root@michael:/build/suricata# clang-format --version
Debian clang-format version 11.0.1-2
root@michael:/build/suricata# git reset HEAD~
Unstaged changes after reset:
M       src/stream-tcp.c
root@michael:/build/suricata# git add src/stream-tcp.c
root@michael:/build/suricata# git clang-format
changed files:
    src/stream-tcp.c
root@michael:/build/suricata# git diff
diff --git a/src/stream-tcp.c b/src/stream-tcp.c
index e3f14f2ac..cf05d498e 100644
--- a/src/stream-tcp.c
+++ b/src/stream-tcp.c
@@ -1638,10 +1638,10 @@ static int StreamTcpPacketStateSynSent(ThreadVars *tv, Packet *p,
                     ssn->client.last_ack);
         } else if (PKT_IS_TOSERVER(p)) {
             /*
-            * On retransmitted SYN packets, the timestamp value must be updated,
-            * to avoid dropping any SYN+ACK packets that respond to a retransmitted SYN
-            * with an updated timestamp in StateSynSentValidateTimestamp.
-            */
+             * On retransmitted SYN packets, the timestamp value must be updated,
+             * to avoid dropping any SYN+ACK packets that respond to a retransmitted SYN
+             * with an updated timestamp in StateSynSentValidateTimestamp.
+             */
             if ((ssn->client.flags & STREAMTCP_STREAM_FLAG_TIMESTAMP) && TCP_HAS_TS(p)) {
                 uint32_t ts_val = TCP_GET_TSVAL(p);

@@ -1651,7 +1651,8 @@ static int StreamTcpPacketStateSynSent(ThreadVars *tv, Packet *p,
                     ssn->client.last_pkt_ts = p->ts.tv_sec;
                 }

-                SCLogDebug("ssn %p: Retransmitted SYN. Updated timestamp from packet %"PRIu64, ssn, p->pcap_cnt);
+                SCLogDebug("ssn %p: Retransmitted SYN. Updated timestamp from packet %" PRIu64, ssn,
+                        p->pcap_cnt);
             }
         }
```

inashivb commented:
> I just received an email that this has failed again; however, the website doesn't reflect that for me.
> 
> Here is what I did including the changes that I made and I pushed them to the branch. Any idea why this isn't happy?
> 
> ```
> root@michael:/build/suricata# clang-format --version
> Debian clang-format version 11.0.1-2
> root@michael:/build/suricata# git reset HEAD~
> Unstaged changes after reset:
> M       src/stream-tcp.c
> root@michael:/build/suricata# git add src/stream-tcp.c
> root@michael:/build/suricata# git clang-format
> changed files:
>     src/stream-tcp.c
> root@michael:/build/suricata# git diff
> diff --git a/src/stream-tcp.c b/src/stream-tcp.c
> index e3f14f2ac..cf05d498e 100644
> --- a/src/stream-tcp.c
> +++ b/src/stream-tcp.c
> @@ -1638,10 +1638,10 @@ static int StreamTcpPacketStateSynSent(ThreadVars *tv, Packet *p,
>                      ssn->client.last_ack);
>          } else if (PKT_IS_TOSERVER(p)) {
>              /*
> -            * On retransmitted SYN packets, the timestamp value must be updated,
> -            * to avoid dropping any SYN+ACK packets that respond to a retransmitted SYN
> -            * with an updated timestamp in StateSynSentValidateTimestamp.
> -            */
> +             * On retransmitted SYN packets, the timestamp value must be updated,
> +             * to avoid dropping any SYN+ACK packets that respond to a retransmitted SYN
> +             * with an updated timestamp in StateSynSentValidateTimestamp.
> +             */
>              if ((ssn->client.flags & STREAMTCP_STREAM_FLAG_TIMESTAMP) && TCP_HAS_TS(p)) {
>                  uint32_t ts_val = TCP_GET_TSVAL(p);
> 
> @@ -1651,7 +1651,8 @@ static int StreamTcpPacketStateSynSent(ThreadVars *tv, Packet *p,
>                      ssn->client.last_pkt_ts = p->ts.tv_sec;
>                  }
> 
> -                SCLogDebug("ssn %p: Retransmitted SYN. Updated timestamp from packet %"PRIu64, ssn, p->pcap_cnt);
> +                SCLogDebug("ssn %p: Retransmitted SYN. Updated timestamp from packet %" PRIu64, ssn,
> +                        p->pcap_cnt);
>              }
>          }
> ```

Seems to me, this should work. Waiting for the CI job to check for sure. Make sure to have committed these changes.

Side note: We prefer "no force push" but "new PR per revision". You can check the contribution guidelines [here](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing)

mtremer commented:
> Side note: We prefer "no force push" but "new PR per revision". You can check the contribution guidelines [here](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing)

I am sorry, I am trying to follow these, but there is so much going on :)

victorjulien commented:
https://redmine.openinfosecfoundation.org/issues/4376

victorjulien commented:
https://github.com/OISF/suricata-verify/pull/849

victorjulien commented:
replaced by #7523
